### PR TITLE
More kubectl multi tenancy

### PR DESCRIPTION
**Pull request by chenqianfzh**
This PR adds multi-tenancy support to the following kubectl commands/sub-commands:
* kubectl edit
* kubectl diff
* kubectl replace
* kubectl apply edit-last-applied
* kubectl apply set-last-applied
* kubectl set image
* kubectl set env
* kubectl set resources
* kubectl set serviceaccount
* kubectl set subject

When this PR is checked in, all the kubectl commands will support multi-tenancy except:
* kubectl certificates
* kubectl auth 

The multi-tenancy work of the commands "kubectl certificates" and "kubectl auth" involve more changes in the server side and will be dealt with another two different PRs. 

### Verification In My Dev Box

Created a deployment based on the file test/yaml/multi_tenancy_examples/tenant1-ns-a-deployment.yaml in the repo and ran the following tests.

#### about "kubectl set image"

 "--tenant" option is in effect now. Without this change, the following command will return "NotFound" error.

```
qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl set image deployment/sample-nginx-deployment nginx=nginx:1.16.1 --record=true --tenant sample-tenant-1 --namespace sample-namespace-a
deployment.extensions/sample-nginx-deployment image updated
```

Output from "kubectl describe deployment" showed the change is in place. The output is omitted as it is too long.

#### about "kubectl set env"

 "--tenant" option is in effect now. Without this change, the following command will return "NotFound" error.
```
qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl set env deployment/sample-nginx-deployment STORAGE_DIR=/local --tenant sample-tenant-1 --namespace sample-namespace-a
deployment.extensions/sample-nginx-deployment env updated

qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl set env deployment/sample-nginx-deployment --list --tenant sample-tenant-1 --namespace sample-namespace-a
# Deployment sample-nginx-deployment, container nginx
STORAGE_DIR=/local
```
#### about "kubectl set resources "

 "--tenant" option is in effect now. Without this change, the following command will return "NotFound" error.

```
qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl set resources deployment/sample-nginx-deployment --limits=cpu=200m --tenant sample-tenant-1 --namespace sample-namespace-a
deployment.extensions/sample-nginx-deployment resource requirements updated
```

Output from "kubectl describe deployment" showed the change is in place. The output is omitted as it is too long.

#### about "kubectl set serviceaccount"

 "--tenant" option is in effect now. Without this change, the following command will return "NotFound" error.
```
qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$  kubectl create serviceaccount test  --tenant sample-tenant-1 --namespace sample-namespace-a  
serviceaccount/test created

qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl set serviceaccount deployment sample-nginx-deployment test  --tenant sample-tenant-1 --namespace sample-namespace-a
deployment.extensions/sample-nginx-deployment serviceaccount updated

qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl create clusterrolebinding my-binding --clusterrole=admin-role --serviceaccount=sample-namespace-a:default --tenant sample-tenant-1
clusterrolebinding.rbac.authorization.k8s.io/my-binding created

qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl describe clusterrolebinding my-binding --tenant sample-tenant-1
Name:         my-binding
Labels:       <none>
Annotations:  <none>
Role:
  Kind:  ClusterRole
  Name:  admin-role
Subjects:
  Kind            Name     Namespace
  ----            ----     ---------
  ServiceAccount  default  sample-namespace-a
```
#### about "kubectl set subject"

 "--tenant" option is in effect now. Without this change, the following command will return "NotFound" error.

```
qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl set subject clusterrolebinding my-binding --serviceaccount=sample-namespace-a:test --tenant sample-tenant-1
clusterrolebinding.rbac.authorization.k8s.io/my-binding subjects updated

qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl describe clusterrolebinding my-binding --tenant sample-tenant-1
Name:         my-binding
Labels:       <none>
Annotations:  <none>
Role:
  Kind:  ClusterRole
  Name:  admin-role
Subjects:
  Kind            Name     Namespace
  ----            ----     ---------
  ServiceAccount  default  sample-namespace-a
  ServiceAccount  test     sample-namespace-a
```

#### about "kubectl set subject"

"--tenant" option is in effect now. Without this change, the following command will return "NotFound" error.

```
qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl replace -f ~/log/replace.yaml --tenant sample-tenant-1 --namespace sample-namespace-a
deployment.apps/sample-nginx-deployment replaced
```

Output from "kubectl describe deployment" showed the change is in place. The output is omitted as it is too long.

#### about "kubectl apply set-last-applied"

"--tenant" option is in effect now. Without this change, the following command will return "NotFound" error.

```
qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl apply set-last-applied -f ~/gopath/qianc/arktos/test/yaml/multi_tenancy_examples/tenant1-ns-a-deployment.yaml --tenant sample-tenant-1 --namespace sample-namespace-a --create-annotation
deployment.apps/sample-nginx-deployment configured
```
Output from "kubectl describe deployment" showed the "last-applied" configuration is added. The output is omitted as it is too long.

"--tenant" option is in effect now. Without this change, the following command will return "NotFound" error.

#### about "kubectl apply edit-last-applied"

"--tenant" option is in effect now. Without this change, the following command will return "NotFound" error.

```
qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl apply edit-last-applied -f ~/gopath/qianc/arktos/test/yaml/multi_tenancy_examples/tenant1-ns-a-deployment.yaml --tenant sample-tenant-1 --namespace sample-namespace-a 
deployment.apps/sample-nginx-deployment edited
```
Output from "kubectl describe deployment" showed the "last-applied" configuration is updated. The output is omitted as it is too long.

#### about "kubectl edit"

"--tenant" option is in effect now. Without this change, the following command will return "NotFound" error.

```
qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl edit deployment sample-nginx-deployment --tenant sample-tenant-1 --namespace sample-namespace-a
deployment.apps/sample-nginx-deployment edited
```

Output from "kubectl describe deployment" showed the deployment is updated. The output is omitted as it is too long.

#### about "kubectl diff"

"--tenant" option is in effect now. Without this change, the following command will return "NotFound" error.

```
qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl diff --tenant sample-tenant-1 --namespace sample-namespace-a -f ~/gopath/qianc/arktos/test/yaml/multi_tenancy_examples/tenant1-ns-a-deployment.yaml
diff -u -N /tmp/LIVE-950103037/apps.v1.Deployment.sample-tenant-1.sample-namespace-a.sample-nginx-deployment /tmp/MERGED-557671224/apps.v1.Deployment.sample-tenant-1.sample-namespace-a.sample-nginx-deployment
--- /tmp/LIVE-950103037/apps.v1.Deployment.sample-tenant-1.sample-namespace-a.sample-nginx-deployment	2020-07-13 09:44:55.247418744 -0700
+++ /tmp/MERGED-557671224/apps.v1.Deployment.sample-tenant-1.sample-namespace-a.sample-nginx-deployment	2020-07-13 09:44:55.255418798 -0700
@@ -6,7 +6,7 @@
     kubectl.kubernetes.io/last-applied-configuration: |
       {"apiVersion":"apps/v1","kind":"Deployment","metadata":{"labels":{"app":"nginx"},"name":"sample-nginx-deployment","namespace":"sample-namespace-a","tenant":"sample-tenant-1"},"spec":{"replicas":2,"selector":{"matchLabels":{"app":"nginx"}},"template":{"metadata":{"labels":{"app":"nginx"}},"spec":{"containers":[{"image":"nginx:1.7.9","name":"nginx","ports":[{"containerPort":80}]}]}}}}
   creationTimestamp: "2020-07-13T16:26:38Z"
-  generation: 8
+  generation: 9
   hashKey: 8348988409592196979
   labels:
     app: nginx
@@ -18,7 +18,7 @@
   uid: 8e9c2aab-3883-41af-af99-0d515e2971f8
 spec:
   progressDeadlineSeconds: 600
-  replicas: 2
+  replicas: 3
   revisionHistoryLimit: 10
   selector:
     matchLabels:
@@ -35,7 +35,7 @@
         app: nginx
     spec:
       containers:
-      - image: nginx:1.7.1
+      - image: nginx:1.7.9
         imagePullPolicy: IfNotPresent
         name: nginx
         ports:
exit status 1

```

### CODE REVIEW

File: pkg/kubectl/cmd/apply/apply_set_last_applied.go [hunk 1]
```diff
@@ -128,6 +130,10 @@ func (o *SetLastAppliedOptions) Complete(f cmdutil.Factory, cmd *cobra.Command)
 	if err != nil {
 		return err
 	}
+	o.tenant, o.enforceTenant, err = f.ToRawKubeConfigLoader().Tenant()
```
Comment (h-w-chen) [inline pkg/kubectl/cmd/apply/apply_set_last_applied.go:133]: Nit: consider use one lines of if ..., err = ...; err!=nil {

File: pkg/kubectl/cmd/apply/apply_set_last_applied.go [hunk 2]
```diff
@@ -151,8 +157,9 @@ func (o *SetLastAppliedOptions) Complete(f cmdutil.Factory, cmd *cobra.Command)
 func (o *SetLastAppliedOptions) Validate() error {
 	r := o.builder.
 		Unstructured().
+		TenantParam(o.tenant).DefaultTenant().
```
Comment (h-w-chen) [inline pkg/kubectl/cmd/apply/apply_set_last_applied.go:160]: does it need to consider both the normal form, and short-path form supporting the tenant related verbs?

File: pkg/kubectl/cmd/apply/apply_set_last_applied.go (hunk 1 above)
Comment (chenqianfzh) [reply pkg/kubectl/cmd/apply/apply_set_last_applied.go:133]: All the code to get commandline params in kubectl part does not use one-lines go get err ( like the code right before). So I am not going to be an outliner. ;-)

File: pkg/kubectl/cmd/apply/apply_set_last_applied.go (hunk 2 above)
Comment (chenqianfzh) [reply pkg/kubectl/cmd/apply/apply_set_last_applied.go:160]: The short-path will be taken care in the apiserver. To use the tenant from the command line or the user context will be taken care by f.ToRawKubeConfigLoader(). So we are good here.

**Pull request merged**